### ISSUE TITLE
Kernel enable_backward=True drops @wp.func gradients in a disabled module

### ISSUE THREAD

**Issue by shi-eric** [NVIDIA/warp#1861]
### Bug Description

When a module has `enable_backward=False`, a kernel with `enable_backward=True` produces a zero gradient if it calls a `@wp.func` helper.

The equivalent calculation written directly in the kernel produces the correct gradient. Warp documents that the more-specific kernel setting takes precedence over the module setting.

#### Reproduction

Save this outside a Warp checkout as `repro.py`:

```python
import warp as wp


wp.set_module_options({"enable_backward": False})


@wp.func
def square(x: float):
    return x * x


@wp.kernel(enable_backward=True)
def square_kernel(x: wp.array[float], y: wp.array[float]):
    y[0] = square(x[0])


x = wp.array([3.0], dtype=float, device="cpu", requires_grad=True)
y = wp.zeros(1, dtype=float, device="cpu", requires_grad=True)

with wp.Tape() as tape:
    wp.launch(square_kernel, dim=1, inputs=[x], outputs=[y], device="cpu")

tape.backward(loss=y)
print(x.grad.numpy())

assert x.grad.numpy()[0] == 6.0
```

Run it with:

```console
uv run --no-project --with warp-lang==1.16 python repro.py
```

The script prints `[0.]` and the assertion fails. The expected gradient is `[6.]`.

For comparison, writing the same calculation directly in the kernel produces `[6.]`:

```python
@wp.kernel(enable_backward=True)
def square_kernel(x: wp.array[float], y: wp.array[float]):
    y[0] = x[0] * x[0]
```

I reproduced this with Warp 1.16.0 and on `main` at commit `808ddbdc0`, reported as Warp 1.18.0.dev0.

#### Expected behavior

A kernel-level `enable_backward=True` should apply when the kernel calls a `@wp.func` helper, even if the module-level setting is `False`.

#### Related issue

This may be related to #942, which tracks compiling adjoint code on demand. This report concerns the incorrect gradient produced by the current configuration behavior.

### System Information

- Warp 1.16.0
- Also reproduced on `main` at `808ddbdc0`, Warp 1.18.0.dev0
- Linux x86_64
- CPU device

**Comment by shi-eric** [comment]
This is addressed by 27df5c23def4c3838f094d0fefaa7ca692c35364.

The change applies the effective kernel-level `enable_backward=True` override throughout the called-function dependency graph, so a module-level `enable_backward=False` no longer causes `@wp.func` adjoints to be omitted for that kernel.

### Public API behavior

- Existing `@wp.kernel(enable_backward=True)` precedence now works as documented when the kernel calls `@wp.func` helpers. No new public API was added.

### Test coverage

- `warp/tests/test_options.py` adds `test_kernel_enable_backward_propagates_to_function`, covering the reported gradient behavior across CPU and CUDA devices.
- `warp/tests/test_codegen.py` adds `TestCodeGen.test_module_enable_backward_controls_function_adjoint_codegen`, verifying that called-function adjoints are emitted when backward generation is enabled and remain omitted when it is disabled.

### Behavioral verification

The issue's CPU reproducer now produces the expected gradient `[6.]`. A transitive `@wp.func` call chain also produced the expected gradient.
